### PR TITLE
Restructure actions and add to reducer

### PR DESCRIPTION
**Pull request by peacebonez**
This PR takes a lot of code that is being reused in various components and is implemented into the userContext actions and the reducer. 

-Errors are now set with the reducer
-Errors are now cleared with the reducer 
-login, signup, logout, clearErrors, and fetchInvites are all now actions in userContext
-I also created an AppAlert component that reads the error message from the reducer and displays it when appropriate

### CODE REVIEW

File: client/src/components/SideBarHeader.js [hunk 1]
```diff
@@ -67,14 +65,7 @@ const SideBarHeader = () => {
...
 
-  const handleLogout = async () => {
-    history.push('/');
-    await axios.get('/user/logout');
-    try {
-    } catch (err) {
-      console.log(err.message);
-    }
-  };
+  if (!userState.user.name) return <Redirect to="/" />;
```
Comment (ThinkTink) [inline client/src/components/SideBarHeader.js:69]: would just userState.user work?

File: client/src/contexts/userContext.js [hunk 2]
```diff
@@ -21,6 +25,33 @@ const UserProvider = (props) => {
   const [userState, dispatch] = useReducer(userReducer, initialState);
 
   const actions = {
+    signUpUser: async (name, email, password, primaryLanguage) => {
+      try {
+        let res = await axios.post('/user/signup', {
```
Comment (ThinkTink) [inline client/src/contexts/userContext.js]: const

File: client/src/contexts/userContext.js [hunk 3]
```diff
@@ -29,20 +60,80 @@ const UserProvider = (props) => {
...
+        if (err.message.includes('404'))
+          dispatch({
+            type: USER_ERROR,
+            payload: { errorMsg: 'User Not Found' },
+          });
+        if (err.message.includes('500'))
+          dispatch({
+            type: USER_ERROR,
+            payload: { errorMsg: 'Server Error' },
+          });
```
Comment (ThinkTink) [inline client/src/contexts/userContext.js]: could set a variable for errorMesg in each if statement and then have just one dispatch at the end

File: client/src/reducers/userReducer.js [hunk 4]
```diff
@@ -1,13 +1,27 @@
...
+          name: '',
+          email: '',
+          avatar: '',
+          contacts: [],
+          primaryLanguage: '',
+        },
+      };
     case USER_ERROR:
-      return { ...state, user: { name: '' } };
+      return { ...state, user: payload };
```
Comment (ThinkTink) [inline client/src/reducers/userReducer.js]: when we dispatch user_error above, it has an error message, not a user in the payload? Is errorMsg a new field on the user object?

**Review by ThinkTink — APPROVED** [review]
Nice work!

File: client/src/components/SideBarHeader.js (hunk 1 above)
Comment (peacebonez) [reply client/src/components/SideBarHeader.js:69]: I don't think so because our init state is: 
```
const initialState = {
  user: { name: '' },
};
```

Comment (peacebonez) [reply client/src/components/SideBarHeader.js:69]: but maybe inititialState should be {}

File: client/src/contexts/userContext.js (hunk 3 above)
Comment (peacebonez) [reply client/src/contexts/userContext.js]: I like that

Comment (peacebonez) [reply client/src/contexts/userContext.js]:
```
 let errorMsg;
        if (err.message.includes('400')) errorMsg = 'Invalid Credentials'   
        if (err.message.includes('404')) errorMsg = 'User not found'
        if (err.message.includes('500')) errorMsg = 'Server error'

          dispatch({
            type: USER_ERROR,
            payload: { errorMsg },
          });
```

File: client/src/components/SideBarHeader.js (hunk 1 above)
Comment (ThinkTink) [reply client/src/components/SideBarHeader.js:69]: I think either way works, we can leave it as is for now

**Pull request merged**